modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexBuildTask.java

```diff
@@ -0,0 +1,223 @@
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed.index;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.concurrent.CompletableFuture.failedFuture;
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+import static java.util.stream.Collectors.toList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.raft.service.RaftGroupService;
+import org.apache.ignite.internal.storage.MvPartitionStorage;
+import org.apache.ignite.internal.storage.RowId;
+import org.apache.ignite.internal.storage.index.IndexStorage;
+import org.apache.ignite.internal.table.distributed.TableMessagesFactory;
+import org.apache.ignite.internal.table.distributed.command.BuildIndexCommand;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.lang.IgniteStringFormatter;
+
+/**
+ * Task of building a table index.
+ */
+class IndexBuildTask {
+    private static final IgniteLogger LOG = Loggers.forClass(IndexBuildTask.class);
+
+    private static final TableMessagesFactory TABLE_MESSAGES_FACTORY = new TableMessagesFactory();
+
+    private final IndexBuildTaskId taskId;
+
+    private final IndexStorage indexStorage;
+
+    private final MvPartitionStorage partitionStorage;
+
+    private final RaftGroupService raftClient;
+
+    private final ExecutorService executor;
+
+    private final IgniteSpinBusyLock busyLock;
+
+    private final int batchSize;
+
+    private final IgniteSpinBusyLock taskBusyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean taskStopGuard = new AtomicBoolean();
+
+    private final CompletableFuture<Void> taskFuture = new CompletableFuture<>();
+
+    IndexBuildTask(
+            IndexBuildTaskId taskId,
+            IndexStorage indexStorage,
+            MvPartitionStorage partitionStorage,
+            RaftGroupService raftClient,
+            ExecutorService executor,
+            IgniteSpinBusyLock busyLock,
+            int batchSize
+    ) {
+        this.taskId = taskId;
+        this.indexStorage = indexStorage;
+        this.partitionStorage = partitionStorage;
+        this.raftClient = raftClient;
+        this.executor = executor;
+        this.busyLock = busyLock;
+        this.batchSize = batchSize;
+    }
+
+    /**
+     * Starts building the index.
+     */
+    void start() {
+        if (!enterBusy()) {
+            taskFuture.complete(null);
+
+            return;
+        }
+
+        LOG.info("Start building the index: [{}]", getCommonIndexInfo());
+
+        try {
+            supplyAsync(this::handleNextBatch, executor)
```

Comment:
Why do you call `supplyAsync` here? So that `createBatchRowIds` are computed on the executor? Can we do that explicitly instead?

Reply:
I would not like to load a thread that will start building the index, so I think it's right to do it in a separate pool.

Reply:
I agree that it is right, I'm proposing to do `runAsync(createBatchRowIds)` inside the `handleNextBatch` method, not here. This will also help to get rid of `Funciton.identity` below

Reply:
In personal correspondence, we decided to leave it as it is.